### PR TITLE
Adds stubs for xdiff 2.0.1

### PR DESCRIPTION
**Pull request by danaketh**
Friend of mine needed this but I've thought others might find it useful as well. I couldn't find real values for `XDIFF_` constants, so I've replaced them with `0`. Feel free to decline/ignore if not up to standards.

### CODE REVIEW

**Review by williamdes — APPROVED** [review]

**Comment by isfedorov** [comment]
@danaketh 
Thank you for the PR! Could you please rebase your branch on master as there was update of cs-fixer in master and currently tests fails on your branch due to old cs-fixer

**Comment by danaketh** [comment]
I hope the issue is resolved now. I'm not sure about the code inspection tho. Should I add the strict type declaration?

**Comment by isfedorov** [comment]
> I hope the issue is resolved now. I'm not sure about the code inspection tho. Should I add the strict type declaration?

@danaketh yep! the issue is solved. Inspection warnings can be ignored in this case.

File: xdiff/xdiff.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
+<?php
+
+/**
+ * Read a size of file created by applying a binary diff
+ * @link https://www.php.net/manual/en/function.xdiff-file-bdiff-size.php
+ * @param string $file The path to the binary patch created by xdiff_string_bdiff() or xdiff_string_rabdiff() functio
```
Comment (isfedorov) [inline xdiff/xdiff.php:6]: Typo in `functio`

File: xdiff/xdiff.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+ * @param string $file The original file.
+ * @param string $patch The unified patch file. It has to be created using xdiff_string_diff(), xdiff_file_diff()
+ * functions or compatible tools.
+ * @param string $dest Path of the resulting file.
+ * @param int $flags Can be either XDIFF_PATCH_NORMAL (default mode, normal patch) or XDIFF_PATCH_REVERSE
+ * (reversed patch). Starting from version 1.5.0, you can also use binary OR to enable XDIFF_PATCH_IGNORESPACE flag.
+ * @return mixed false if an internal error happened, string with rejected chunks if patch couldn't be applied
+ * or true if patch has been successfully applied.
+ */
+function xdiff_file_patch(string $file, string $patch, string $dest, int $flags = XDIFF_PATCH_NORMAL) {}
```
Comment (isfedorov) [inline xdiff/xdiff.php:95]: You can add attribute `#[ExpectedValues]` for `$flags` parameter here like `xdiff_file_patch(string $file, string $patch, string $dest, #[ExpectedValues([XDIFF_PATCH_NORMAL|XDIFF_PATCH_REVERSE])] int $flags = XDIFF_PATCH_NORMAL)`. Thus these possible values will be suggested in completion

File: xdiff/xdiff.php [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+/**
+ * Patch a file with an unified diff
+ * @link https://www.php.net/manual/en/function.xdiff-file-patch.php
+ * @param string $file The original file.
+ * @param string $patch The unified patch file. It has to be created using xdiff_string_diff(), xdiff_file_diff()
+ * functions or compatible tools.
+ * @param string $dest Path of the resulting file.
+ * @param int $flags Can be either XDIFF_PATCH_NORMAL (default mode, normal patch) or XDIFF_PATCH_REVERSE
+ * (reversed patch). Starting from version 1.5.0, you can also use binary OR to enable XDIFF_PATCH_IGNORESPACE flag.
+ * @return mixed false if an internal error happened, string with rejected chunks if patch couldn't be applied
```
Comment (isfedorov) [inline xdiff/xdiff.php:92]: It makes sense to change `mixed` to `bool|string` to make type inference and inspections more correct and strict

File: xdiff/xdiff.php [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+ * @return int the size of file that would be created.
+ */
+function xdiff_string_bdiff_size(string $patch): int {}
+
+/**
+ * Make binary diff of two strings
+ * @link https://www.php.net/manual/en/function.xdiff-string-bdiff.php
+ * @param string $old_data First string with binary data. It acts as "old" data.
+ * @param string $new_data Second string with binary data. It acts as "new" data.
+ * @return string string with binary diff containing differences between "old" and "new" data or false
```
Comment (isfedorov) [inline xdiff/xdiff.php:121]: Return type should be `string|false` as `false` is return if error occured

File: xdiff/xdiff.php [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+): bool {}
+
+/**
+ * Merge 3 files into one
+ * @link https://www.php.net/manual/en/function.xdiff-file-merge3.php
+ * @param string $old_file Path to the first file. It acts as "old" file.
+ * @param string $new_file1 Path to the second file. It acts as modified version of old_file.
+ * @param string $new_file2 Path to the third file. It acts as modified version of old_file.
+ * @param string $dest Path of the resulting file, containing merged changed from both new_file1 and new_file2.
+ * @return mixed true if merge was successful, string with rejected chunks if it was not or false if an internal
```
Comment (isfedorov) [inline xdiff/xdiff.php:68]: Return type can be changed to `bool|string`

File: xdiff/xdiff.php [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+ * if an internal error occurred.
+ */
+function xdiff_string_bdiff(string $old_data, string $new_data): string {}
+
+/**
+ * Patch a string with a binary diff
+ * @link https://www.php.net/manual/en/function.xdiff-string-bpatch.php
+ * @param string $str The original binary string.
+ * @param string $patch The binary patch string.
+ * @return string the patched string, or false on error.
```
Comment (isfedorov) [inline xdiff/xdiff.php:131]: Return type should be `string|false`

File: xdiff/xdiff.php [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+ */
+function xdiff_string_bpatch(string $str, string $patch): string {}
+
+/**
+ * Alias of xdiff_string_bdiff
+ * @link https://www.php.net/manual/en/function.xdiff-string-diff-binary.php
+ * @param string $old_data First string with binary data. It acts as "old" data.
+ * @param string $new_data Second string with binary data. It acts as "new" data.
+ * @return string string with result or false if an internal error happened.
```
Comment (isfedorov) [inline xdiff/xdiff.php:140]: Return type `string|false`

File: xdiff/xdiff.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+/**
+ * Make unified diff of two strings
+ * @link https://www.php.net/manual/en/function.xdiff-string-diff.php
+ * @param string $old_data First string with data. It acts as "old" data.
+ * @param string $new_data Second string with data. It acts as "new" data.
+ * @param int $context Indicates how many lines of context you want to include in the diff result.
+ * @param bool $minimal Set this parameter to true if you want to minimalize the size of the result
+ * (can take a long time).
+ * @return string string with resulting diff or false if an internal error happened.
```
Comment (isfedorov) [inline xdiff/xdiff.php:152]: The same as above

File: xdiff/xdiff.php [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+function xdiff_string_diff(string $old_data, string $new_data, int $context = 3, bool $minimal = false): string {}
+
+/**
+ * Merge 3 strings into one
+ * @link https://www.php.net/manual/en/function.xdiff-string-merge3.php
+ * @param string $old_data First string with data. It acts as "old" data.
+ * @param string $new_data1 Second string with data. It acts as modified version of old_data.
+ * @param string $new_data2 Third string with data. It acts as modified version of old_data.
+ * @param ?string $error If provided then rejected parts are stored inside this variable.
+ * @return mixed the merged string, false if an internal error happened, or true if merged string is empty.
```
Comment (isfedorov) [inline xdiff/xdiff.php:163]: Return type `string|bool`

File: xdiff/xdiff.php [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+ */
+function xdiff_string_merge3(string $old_data, string $new_data1, string $new_data2, ?string &$error) {}
+
+/**
+ * Alias of xdiff_string_bpatch
+ * @link https://www.php.net/manual/en/function.xdiff-string-patch-binary.php
+ * @param string $str The original binary string.
+ * @param string $patch The binary patch string.
+ * @return string the patched string, or false on error.
```
Comment (isfedorov) [inline xdiff/xdiff.php:172]: Return type `string|false`

File: xdiff/xdiff.php [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+ */
+function xdiff_string_patch(string $str, string $patch, ?int $flags, ?string &$error): string {}
+
+/**
+ * Make binary diff of two strings using the Rabin's polynomial fingerprinting algorithm
+ * @link https://www.php.net/manual/en/function.xdiff-string-rabdiff.php
+ * @param string $old_data First string with binary data. It acts as "old" data.
+ * @param string $new_data Second string with binary data. It acts as "new" data.
+ * @return string string with binary diff containing differences between "old" and "new" data or false if an internal error occurred.
```
Comment (isfedorov) [inline xdiff/xdiff.php:194]: `string|false`

**Comment by isfedorov** [comment]
fixed in other PR

**Pull request closed without merging**